Enable reverse view_funcs by default for python subclasses

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113280
* #113279
* #116308
* __->__ #116512
* #115894

Part 3 of implementation for general [subclass view fake-ification](https://docs.google.com/document/d/1C5taWiplmX7nKiURXDOAZG2W5VNJ2iV0fQFq92H0Cxw).

Changes codegen to generate `view_func()` / `rev_view_func()` by default for python subclasses. With `view_func()` existing more often now, the lazy view rebase logic [here](https://github.com/pytorch/pytorch/blob/f10c3f41848fa2889f75ae13de9cafb0c29bf33e/torch/csrc/autograd/variable.cpp#L665-L695) causes some slight behavior changes for in-place ops on views:
* Additional view nodes are inserted into output graphs, changing their string representation, although they are functionally the same. The extra nodes are removed in AOTAutograd's DCE pass.
* When `t` is a `FunctionalTensor`, calling `t.grad_fn` will now invoke `view_func()`; we need to make sure we're operating in a `FunctionalTensorMode` so the view op calls succeed.

jbschlosser committed:
Enable reverse view_funcs by default for python subclasses

[ghstack-poisoned]

jbschlosser committed:
Update on "Enable reverse view_funcs by default for python subclasses"


Part 3 of implementation for general [subclass view fake-ification](https://docs.google.com/document/d/1C5taWiplmX7nKiURXDOAZG2W5VNJ2iV0fQFq92H0Cxw).

Changes codegen to generate `view_func()` / `rev_view_func()` by default for python subclasses. This has the effect of inserting additional views nodes into output graphs, changing their string representation, although they are functionally the same (see [here](https://github.com/pytorch/pytorch/blob/f10c3f41848fa2889f75ae13de9cafb0c29bf33e/torch/csrc/autograd/variable.cpp#L665-L695) for the call to the now-existing `view_func()`). The extra nodes are removed in AOTAutograd's DCE pass.

[ghstack-poisoned]

jbschlosser committed:
Update on "Enable reverse view_funcs by default for python subclasses"


Part 3 of implementation for general [subclass view fake-ification](https://docs.google.com/document/d/1C5taWiplmX7nKiURXDOAZG2W5VNJ2iV0fQFq92H0Cxw).

Changes codegen to generate `view_func()` / `rev_view_func()` by default for python subclasses. This has the effect of inserting additional views nodes into output graphs, changing their string representation, although they are functionally the same (see [here](https://github.com/pytorch/pytorch/blob/f10c3f41848fa2889f75ae13de9cafb0c29bf33e/torch/csrc/autograd/variable.cpp#L665-L695) for the call to the now-existing `view_func()`). The extra nodes are removed in AOTAutograd's DCE pass.

[ghstack-poisoned]

jbschlosser committed:
Update on "Enable reverse view_funcs by default for python subclasses"


Part 3 of implementation for general [subclass view fake-ification](https://docs.google.com/document/d/1C5taWiplmX7nKiURXDOAZG2W5VNJ2iV0fQFq92H0Cxw).

Changes codegen to generate `view_func()` / `rev_view_func()` by default for python subclasses. This has the effect of inserting additional views nodes into output graphs, changing their string representation, although they are functionally the same (see [here](https://github.com/pytorch/pytorch/blob/f10c3f41848fa2889f75ae13de9cafb0c29bf33e/torch/csrc/autograd/variable.cpp#L665-L695) for the call to the now-existing `view_func()`). The extra nodes are removed in AOTAutograd's DCE pass.

[ghstack-poisoned]

jbschlosser committed:
Update on "Enable reverse view_funcs by default for python subclasses"


Part 3 of implementation for general [subclass view fake-ification](https://docs.google.com/document/d/1C5taWiplmX7nKiURXDOAZG2W5VNJ2iV0fQFq92H0Cxw).

Changes codegen to generate `view_func()` / `rev_view_func()` by default for python subclasses. With `view_func()` existing more often now, the lazy view rebase logic [here](https://github.com/pytorch/pytorch/blob/f10c3f41848fa2889f75ae13de9cafb0c29bf33e/torch/csrc/autograd/variable.cpp#L665-L695) causes some slight behavior changes for in-place ops on views:
* Additional view nodes are inserted into output graphs, changing their string representation, although they are functionally the same. The extra nodes are removed in AOTAutograd's DCE pass.
* When `t` is a `FunctionalTensor`, calling `t.grad_fn` will now invoke `view_func()`; we need to make sure we're operating in a `FunctionalTensorMode` so the view op calls succeed.

[ghstack-poisoned]

jbschlosser committed:
Update on "Enable reverse view_funcs by default for python subclasses"


Part 3 of implementation for general [subclass view fake-ification](https://docs.google.com/document/d/1C5taWiplmX7nKiURXDOAZG2W5VNJ2iV0fQFq92H0Cxw).

Changes codegen to generate `view_func()` / `rev_view_func()` by default for python subclasses. With `view_func()` existing more often now, the lazy view rebase logic [here](https://github.com/pytorch/pytorch/blob/f10c3f41848fa2889f75ae13de9cafb0c29bf33e/torch/csrc/autograd/variable.cpp#L665-L695) causes some slight behavior changes for in-place ops on views:
* Additional view nodes are inserted into output graphs, changing their string representation, although they are functionally the same. The extra nodes are removed in AOTAutograd's DCE pass.
* When `t` is a `FunctionalTensor`, calling `t.grad_fn` will now invoke `view_func()`; we need to make sure we're operating in a `FunctionalTensorMode` so the view op calls succeed.

[ghstack-poisoned]

bdhirsh commented on torch/_functorch/_aot_autograd/traced_function_transforms.py:
```diff
@@ -247,7 +247,8 @@ def inner_fn(primals: List[Any], tangents: List[Any]):
     def inner_fn_with_anomaly(*args):
-        with fx_traceback.preserve_node_meta(), warnings.catch_warnings():
+        with fx_traceback.preserve_node_meta(), warnings.catch_warnings(), \
+                FunctionalTensorMode(aot_config.pre_dispatch):
```
Since we need the `FunctionalTensorMode` in two places - do you mind adding a quick `Note [Accessing .grad_fn on FunctionalTensor]` comment and mentioning it both places so we have a nice paper trail?

jbschlosser replied:
Yep sure thing!

bdhirsh reviewed (approved):
lgtm!

jbschlosser committed:
Update on "Enable reverse view_funcs by default for python subclasses"


Part 3 of implementation for general [subclass view fake-ification](https://docs.google.com/document/d/1C5taWiplmX7nKiURXDOAZG2W5VNJ2iV0fQFq92H0Cxw).

Changes codegen to generate `view_func()` / `rev_view_func()` by default for python subclasses. With `view_func()` existing more often now, the lazy view rebase logic [here](https://github.com/pytorch/pytorch/blob/f10c3f41848fa2889f75ae13de9cafb0c29bf33e/torch/csrc/autograd/variable.cpp#L665-L695) causes some slight behavior changes for in-place ops on views:
* Additional view nodes are inserted into output graphs, changing their string representation, although they are functionally the same. The extra nodes are removed in AOTAutograd's DCE pass.
* When `t` is a `FunctionalTensor`, calling `t.grad_fn` will now invoke `view_func()`; we need to make sure we're operating in a `FunctionalTensorMode` so the view op calls succeed.

[ghstack-poisoned]

jbschlosser committed:
Update on "Enable reverse view_funcs by default for python subclasses"


Part 3 of implementation for general [subclass view fake-ification](https://docs.google.com/document/d/1C5taWiplmX7nKiURXDOAZG2W5VNJ2iV0fQFq92H0Cxw).

Changes codegen to generate `view_func()` / `rev_view_func()` by default for python subclasses. With `view_func()` existing more often now, the lazy view rebase logic [here](https://github.com/pytorch/pytorch/blob/f10c3f41848fa2889f75ae13de9cafb0c29bf33e/torch/csrc/autograd/variable.cpp#L665-L695) causes some slight behavior changes for in-place ops on views:
* Additional view nodes are inserted into output graphs, changing their string representation, although they are functionally the same. The extra nodes are removed in AOTAutograd's DCE pass.
* When `t` is a `FunctionalTensor`, calling `t.grad_fn` will now invoke `view_func()`; we need to make sure we're operating in a `FunctionalTensorMode` so the view op calls succeed.

[ghstack-poisoned]

jbschlosser committed:
Update on "Enable reverse view_funcs by default for python subclasses"


Part 3 of implementation for general [subclass view fake-ification](https://docs.google.com/document/d/1C5taWiplmX7nKiURXDOAZG2W5VNJ2iV0fQFq92H0Cxw).

Changes codegen to generate `view_func()` / `rev_view_func()` by default for python subclasses. With `view_func()` existing more often now, the lazy view rebase logic [here](https://github.com/pytorch/pytorch/blob/f10c3f41848fa2889f75ae13de9cafb0c29bf33e/torch/csrc/autograd/variable.cpp#L665-L695) causes some slight behavior changes for in-place ops on views:
* Additional view nodes are inserted into output graphs, changing their string representation, although they are functionally the same. The extra nodes are removed in AOTAutograd's DCE pass.
* When `t` is a `FunctionalTensor`, calling `t.grad_fn` will now invoke `view_func()`; we need to make sure we're operating in a `FunctionalTensorMode` so the view op calls succeed.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge -f "spurious distributed failure"
